### PR TITLE
[Enhancement] Drop special tokens in kv blending

### LINKED ISSUE #190: [KV Blending] Make hyper-parameters configurable (e.g., recompute ratio)
The current KV blending module can only be controlled by `enable_blending` option in the configuration yaml.

However, multiple internal hyper-parameters, such as the recomputation ratio, could also impact the KV blending module.

Specifically, we need at least 2 more hyper-parameters:
- Recomputation ratio -- controls how much tokens' KV will be re-computed by the KV blending module.
- Minimum # of tokens to enable blending -- if the request is too short, we don't want to trigger KV blending logic.

This feature also requires changing the `blend_adapter.py` in the lmcache_vllm repo. See the code here:
https://github.com/LMCache/lmcache-vllm/blob/b890544e7efbaa2738902889c6f4ac56cc22ecd5/lmcache_vllm/blend_adapter.py#L19

------------
**Update:**

We also need to support changing special tokens

### LINKED ISSUE #194: [KV Blending] Mitigate the impact of separator tokens (currently it generates “# #” in the output)

### PR DESCRIPTION
**Pull request by XbzOnGit**
Drop special tokens, making them only used to separate prompt into segments but not seen by underlying serving system. 

This PR resolves #194 

drop_spt_and_get_indices should be used by tokenizers to remove the special tokens, and record the indices of segments.

new_request now accepts full prompts instead of input_ids(which can cause problem if the prefill of this request does not start on 0, like a vllm internal prefix cache hit or chunked prefill).
It also accepts an external indices because the special tokens should have been removed.

### CODE REVIEW

**Comment by XbzOnGit** [comment]
Here are some logging of blend_kv.py example. Can see that the "#" tokens are not generated now(check "Newly generated text" row).  
```
-------------- Pre-computing KV cache for the chunks -------------------
Processed prompts:   0%|                                                                                                          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]INFO LMCache: KV cache retrieving mode: RetrieveStatus.PREFILL [2024-11-07 20:27:48,420] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_injection.py:48
INFO LMCache: Using default batched implementation of the get() method [2024-11-07 20:27:48,426] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:102
INFO LMCache: Retrieved 0 chunks [2024-11-07 20:27:48,426] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:381
DEBUG LMCache: Returning the original input! [2024-11-07 20:27:48,426] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_adapter.py:648
DEBUG LMCache: Split input tokens into 1 requests [2024-11-07 20:27:48,427] -- /root/autodl-tmp/LMCache/lmcache/blend/retriever.py:290
INFO LMCache: Using default batched implementation of the get() method [2024-11-07 20:27:48,428] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:102
INFO LMCache: Retrieved 0 chunks [2024-11-07 20:27:48,428] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:381
INFO LMCache: KV cache saving mode: [<StoreStatus.PREFILL: 1>] [2024-11-07 20:27:48,522] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_injection.py:127
INFO LMCache: Using default batched implementation of the put() method [2024-11-07 20:27:48,558] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:81
INFO LMCache: Stored/updated 3 chunks, total time 0.01s, make chunks time 0.01s [2024-11-07 20:27:48,558] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:311
DEBUG LMCache: Store skips 0 tokens and then stores 768 tokens [2024-11-07 20:27:48,559] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_adapter.py:474
Processed prompts: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  4.77it/s, est. speed input: 3734.55 toks/s, output: 4.86 toks/s]
Processed prompts:   0%|                                                                                                          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]INFO LMCache: KV cache retrieving mode: RetrieveStatus.PREFILL [2024-11-07 20:27:48,674] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_injection.py:48
INFO LMCache: Using default batched implementation of the get() method [2024-11-07 20:27:48,675] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:102
INFO LMCache: Retrieved 0 chunks [2024-11-07 20:27:48,675] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:381
DEBUG LMCache: Returning the original input! [2024-11-07 20:27:48,675] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_adapter.py:648
DEBUG LMCache: Split input tokens into 1 requests [2024-11-07 20:27:48,675] -- /root/autodl-tmp/LMCache/lmcache/blend/retriever.py:290
INFO LMCache: Using default batched implementation of the get() method [2024-11-07 20:27:48,676] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:102
INFO LMCache: Retrieved 0 chunks [2024-11-07 20:27:48,676] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:381
INFO LMCache: KV cache saving mode: [<StoreStatus.PREFILL: 1>] [2024-11-07 20:27:48,841] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_injection.py:127
INFO LMCache: Using default batched implementation of the put() method [2024-11-07 20:27:48,880] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:81
INFO LMCache: Stored/updated 6 chunks, total time 0.00s, make chunks time 0.00s [2024-11-07 20:27:48,881] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:311
DEBUG LMCache: Store skips 0 tokens and then stores 1520 tokens [2024-11-07 20:27:48,881] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_adapter.py:474
Processed prompts: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  3.73it/s, est. speed input: 5681.97 toks/s, output: 3.74 toks/s]
Running the real query here!
Processed prompts:   0%|                                                                                                          | 0/1 [00:00<?, ?it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]INFO LMCache: KV cache retrieving mode: RetrieveStatus.PREFILL [2024-11-07 20:27:51,947] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_injection.py:48
INFO LMCache: Using default batched implementation of the get() method [2024-11-07 20:27:51,948] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:102
INFO LMCache: Retrieved 0 chunks [2024-11-07 20:27:51,948] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:381
DEBUG LMCache: Returning the original input! [2024-11-07 20:27:51,948] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_adapter.py:648
DEBUG LMCache: Split input tokens into 4 requests [2024-11-07 20:27:51,948] -- /root/autodl-tmp/LMCache/lmcache/blend/retriever.py:290
INFO LMCache: Using default batched implementation of the get() method [2024-11-07 20:27:51,949] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:102
INFO LMCache: Retrieved 0 chunks [2024-11-07 20:27:51,949] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:381
INFO LMCache: Using default batched implementation of the get() method [2024-11-07 20:27:51,950] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:102
INFO LMCache: Retrieved 3 chunks (768 tokens in total) --elapsed time 0.008957004640251398 [2024-11-07 20:27:51,959] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:407
INFO LMCache: Using default batched implementation of the get() method [2024-11-07 20:27:51,959] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:102
INFO LMCache: Retrieved 6 chunks (1520 tokens in total) --elapsed time 0.017295774072408676 [2024-11-07 20:27:51,976] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:407
INFO LMCache: Using default batched implementation of the get() method [2024-11-07 20:27:51,977] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/abstract_backend.py:102
INFO LMCache: Retrieved 0 chunks [2024-11-07 20:27:51,977] -- /root/autodl-tmp/LMCache/lmcache/cache_engine.py:381
INFO LMCache: Selected 368 tokens out of 2313 tokens to blend [2024-11-07 20:27:52,045] -- /root/autodl-tmp/LMCache/lmcache/blend/executor.py:178
DEBUG LMCache: Updating selected_token_indices to tensor([367], device='cuda:0', dtype=torch.int32) after blending [2024-11-07 20:27:52,136] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_injection.py:139
Processed prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  1.42it/s, est. speed input: 3275.61 toks/s, output: 42.48 toks/s]
Newly generated text: '\n\n       This document mainly talks about the process of transcoding video and audio files using the FFmpeg multimedia framework. It explains the'
DEBUG LMCache: Closing LMCache Engine [2024-11-07 20:27:52,649] -- /root/autodl-tmp/lmcache-vllm/lmcache_vllm/vllm_adapter.py:237
INFO LMCache: Closed the put worker in local backend [2024-11-07 20:27:52,650] -- /root/autodl-tmp/LMCache/lmcache/storage_backend/local_backend.py:208
```

File: lmcache/blend/interfaces.py [hunk 1]
```diff
@@ -76,17 +76,16 @@ class BlendRetriever(metaclass=abc.ABCMeta):
...
         """Create a new BlendRetrieverTask to retrieve the KV caches.
         It may launch async tasks in the background during the retrieval.
 
-        :param torch.Tensor input_tokens: The input tokens, could include
-            multiple requests in a batch
-        :param torch.Tensor query_start_loc: The start location of the query if
-            input_tokens has multiple requests in a batch. The length should be
-            the number of requests in the batch + 1
+        :param List[torch.Tensor] full_prompts: The full prompts for each
+        request in this batch.
```
Comment (ApostaC) [inline lmcache/blend/interfaces.py:86]: Can you add the following sentence:", which will contain the tokens hitting the vLLM's internal prefix caching" (hope my understanding is correct)

File: lmcache/blend/retriever.py [hunk 2]
```diff
@@ -187,10 +187,10 @@ class SPTBlendRetriever(BlendRetriever):
     of the input text chunk.
 
     Example:
-        Input = [x, x, x, spt, y, y, spt, z, z, z, z]
```
Comment (ApostaC) [inline lmcache/blend/retriever.py:190]: Could you keep this line also? Maybe it is deleted accidentally?

File: lmcache/blend/retriever.py [hunk 3]
```diff
@@ -214,60 +214,73 @@ def __init__(
...
+        while True:
+            next_len = i + spt_len
+            if next_len > len(full_prompt):
+                break
+            if full_prompt[i:next_len] == self.spt:
+                splitted_tokens.append(full_prompt[start:i])
+                start = next_len
+                i = next_len
+            else:
+                i += 1
```
Comment (ApostaC) [inline lmcache/blend/retriever.py:245]: Could you please use the tensor operation to do this? Python loops could be slow when the input tokens are long. (You can take a look at the previous implementation)

File: lmcache/blend/retriever.py [hunk 4]
```diff
@@ -214,60 +214,73 @@ def __init__(
...
+            new_indices.append(new_indices[-1] + len(split))
+        new_indices[-1] = len(new_prompt)
+        return new_prompt, new_indices
 
     def new_request(
         self,
-        input_tokens: torch.Tensor,
-        query_start_loc: torch.Tensor,
+        full_prompts: List[torch.Tensor],
+        indices: List[List[int]],
```
Comment (YaoJiayi) [inline lmcache/blend/retriever.py:264]: Why are we changing everything from tensor to list?

File: lmcache/blend/retriever.py [hunk 5]
```diff
@@ -214,60 +214,73 @@ def __init__(
...
-
+            for prompt_idx, prompt in enumerate(full_prompts):
+                prompt_indices = indices[prompt_idx]
+                assert len(prompt_indices) >= 2
+                segment_num = len(prompt_indices) - 1
+                for i in range(segment_num):
+                    start_loc = prompt_indices[i]
+                    end_loc = prompt_indices[i + 1]
+                    if end_loc > start_loc:
+                        splitted_tokens.append(prompt[start_loc:end_loc])
```
Comment (YaoJiayi) [inline lmcache/blend/retriever.py]: Same as above. Can we use as much tensor operations as possible?

**Review by YaoJiayi — COMMENTED** [review]
Thanks Sixian! Left some comments in both this and the lmcache-vllm pr.

**Review by ApostaC — CHANGES_REQUESTED** [review]
Thanks for the great work, Sixian! Two minor comments + 1 potential improvement. Please see the details below:

File: lmcache/blend/retriever.py (hunk 3 above)
Comment (XbzOnGit) [reply lmcache/blend/retriever.py:245]: > Could you please use the tensor operation to do this? Python loops could be slow when the input tokens are long. (You can take a look at the previous implementation)

A potential problem is tokenizer is returning a list of ints. So if we want to do it in tensor, we have to construct a tensor from list, do it in tensor, and convert it back to list.
So not sure if it is still an improvement in this case.

File: lmcache/blend/retriever.py (hunk 4 above)
Comment (XbzOnGit) [reply lmcache/blend/retriever.py:264]: > Why are we changing everything from tensor to list?

For full_prompts, they are just tensors from seq_group_list. So they are organized in a list by vllm before. And if we want to make it one tensor, we need to explicitly concatenate it. I think it is unnecessary. (Note that we cannot use input_ids, the one tensor before for it drops vllm internal cache hit tokens and chunked prefill tokens, if we want that compatibility in the future)

For indices, it is possible to make it List[tensor], but I was just thinking we will probably not cut one prompt into thousands of text chunks.

Comment (XbzOnGit) [reply lmcache/blend/retriever.py:264]: Also, I think even if it is input_tokens, we still loop on the query_start_loc, so the List in full_prompts is not an overhead comparing with before.

File: lmcache/blend/retriever.py (hunk 5 above)
Comment (XbzOnGit) [reply lmcache/blend/retriever.py]: > Same as above. Can we use as much tensor operations as possible?

The indices part is possible to improve, but I think we will not cut it into too many chunks, probably in 10 in one prompt.

File: lmcache/blend/retriever.py (hunk 3 above)
Comment (ApostaC) [reply lmcache/blend/retriever.py:245]: Can you write a simple offline benchmark script to test? For example, create a list of 25K integers, run two splitting functions for 100 rounds, and see which one is faster.
Don't need to be too complicated. Thanks!

Comment (ApostaC) [reply lmcache/blend/retriever.py:245]: Just to archive the testing results here. Both tensor implementation and list implementation has similar performance on the server (around 4ms).

Code for testing:
```python
import torch
from typing import List, Tuple
import time
import random
import sys
class DropClass:
    def __init__(self, spt: List[int]):
         self.spt = spt
         self.tensor_spt = torch.tensor(spt, device="cpu")
    def drop_spt_and_get_indices(
                self, full_prompt: List[int]) -> Tuple[List[int], List[int]]:
        spt_len = len(self.spt)
        assert spt_len >= 1
        i = 0
        splitted_tokens = []
        start = 0
        while True:
            next_len = i + spt_len
            if next_len > len(full_prompt):
                break
            if full_prompt[i:next_len] == self.spt:
                splitted_tokens.append(full_prompt[start:i])
                start = next_len
                i = next_len
            else:
                i += 1

        if start < len(full_prompt):
            splitted_tokens.append(full_prompt[start:])

        new_prompt = []
        new_indices = []
        this_seg_start = 0
        for split in splitted_tokens:
            new_prompt.extend(split)
            new_indices.append(this_seg_start + len(split))
            this_seg_start = new_indices[-1]
        if len(new_indices) > 0:
            new_indices.pop()
        return new_prompt, new_indices

    def drop_tensor_edition(self, full_prompt: List[int]) -> Tuple[List[int], torch.Tensor]:
        spt_len = len(self.tensor_spt)
        assert spt_len >= 1
        prompt_tensor = torch.tensor(full_prompt, device="cpu")
        # TODO: Drop and get indices.
        # Make an efficient tensor operation.
        if spt_len == 1:
            indices = (prompt_tensor == self.tensor_spt).nonzero().squeeze()
        else:
            windows = prompt_tensor.unfold(0, spt_len, 1)
            indices = (windows == self.tensor_spt).all(dim=1).nonzero().squeeze()
        if indices.dim() == 0:
            indices = indices.unsqueeze(0)
        start = 0
        splitted_tokens = []
        new_indices = []
        for i in indices:
            splitted_tokens.append(prompt_tensor[start:i])
            start = i + spt_len
        if start < len(prompt_tensor):
            splitted_tokens.append(prompt_tensor[start:])
        new_indices = []
        this_seg_start = 0
        for split in splitted_tokens:
            new_indices.append(this_seg_start + len(split))
            this_seg_start = new_indices[-1]
        if len(new_indices) > 0:
            new_indices.pop()
        new_indices = torch.tensor(new_indices, device="cpu")
        new_prompt_tensor = torch.cat(splitted_tokens, dim=0)
        new_prompt = new_prompt_tensor.tolist()
        return new_prompt, new_indices

        


if __name__ == "__main__":
    if len(sys.argv) != 5:
        print(f"Usage: python3 {sys.argv[0]} <prompt_length_without_spt> <indices_length> <spt_length> <repeat_times>", file=sys.stderr)
        sys.exit(1)
    prompt_length = int(sys.argv[1])
    indices_length = int(sys.argv[2])
    spt_length = int(sys.argv[3])
    repeat_times = int(sys.argv[4])
    assert repeat_times >= 1
    assert prompt_length >= 1
    assert spt_length >= 1
    spt = [10010] * spt_length
    prompt = [random.randint(0, 10000) for _ in range(prompt_length)]
    segment_num = indices_length + 1
    seg_len = prompt_length // segment_num
    indices = [seg_len * i for i in range(1, segment_num)]
    if len(indices) > 0 and indices[-1] == prompt_length:
        indices.pop()
    new_prompt = []
    this_seg_start = 0
    for i in indices:
        new_prompt.extend(prompt[this_seg_start:i])
        new_prompt.extend(spt)
        this_seg_start = i
    if this_seg_start < prompt_length:
        new_prompt.extend(prompt[this_seg_start:])
    drop_class = DropClass(spt)
    # Check correctness.
    ans_prompt, ans_indices = drop_class.drop_spt_and_get_indices(new_prompt)
    assert ans_prompt == prompt
    assert ans_indices == indices, f"{ans_indices} != {indices}"
    ans_prompt, ans_indices = drop_class.drop_tensor_edition(new_prompt)
    assert ans_prompt == prompt
    assert ans_indices.tolist() == indices, f"{ans_indices.tolist()} != {indices}"
    # Benchmark.
    start_time = time.perf_counter()
    for _ in range(repeat_times):
        ans_p, ans_i = drop_class.drop_spt_and_get_indices(new_prompt)
    end_time = time.perf_counter()
    print(f"avg time for list edition:   {(end_time - start_time) / repeat_times}")
    start_time = time.perf_counter()
    for _ in range(repeat_times):
        ans_p, ans_i = drop_class.drop_tensor_edition(new_prompt)
    end_time = time.perf_counter()
    print(f"avg time for tensor edition: {(end_time - start_time) / repeat_times}")
```

**Review by ApostaC — APPROVED** [review]
LGTM! @YaoJiayi Please let us know if you have any last minute comments, thanks!

**Comment by ApostaC** [comment]
[blend_performance_test.zip](https://github.com/user-attachments/files/17718811/blend_performance_test.zip)
Also attaching the testing results here (credit to @XbzOnGit ):
```
TTFT:
no cache --> 6.3484726920723915 second
vllm internal cache --> 4.7106399312615395 second
LMCache without blending --> 5.070993646979332 second
LMCache with blending --> 1.8784587979316711 second
```

**Comment by YaoJiayi** [comment]
Look good to me. Let’s merge it after resolving the conflict.

**Pull request merged**